### ISSUE TITLE
FR: Kubernetes operator to provide egress proxy

### ISSUE THREAD

**Issue by parasol-me** [tailscale/tailscale#8184]
### What are you trying to do?

Our stack consists of two Kubernetes clusters, each with an API service.
The backing of the API service is a deployment to scale the pod replicas.
We want to provide two-way connection between these services over our tailnet in order to avoid public ingress and authentication.

We have been using the tailscale operator to provide inbound connectivity to the services, but in order to provide outbound connectivity to the API pods we have the following options:

- Manually create a tailscale proxy service in each cluster that points to the IP address of the service exposed by operator from the other cluster
- Manually add tailscale sidecar to the api pods

We are deploying many of these two cluster setups, so managing the hostnames for these services is important.

The first option is a bit limited because it requires providing the destination IP address and port.

For the second option, the sidecar already provides two-way communication, but we don't actually want each pod in the deployment to be exposed as a host. It should only provide outbound connection to the service exposed by operator.

### How should we solve this?

It would be good if the operator could provide both outbound and inbound connections via either an additional proxy (ideally only hostname required) or by attaching sidecars to pods but with only an outbound connection (not exposing each pod as a tailnet host).

Maybe there is a better way to achieve it, we are also reaching out for suggestions.

Basically, we would like to use only the operator to provide bi-directional communication between services where the backing is a scalable deployment.

### What is the impact of not solving this?

Currently, we are either manually creating outbound proxies to specific IP addresses or manually attaching a sidecar to all pods but only for outbound connections.

In combination with using the operator to tag services for inbound connection.

### Anything else?

N/A

**Comment by maisem** [comment]
We have been thinking about this problem a bit, would you need Tailscale provisioned TLS certs in this setup? I ask because there are a few ways to expose the service. Either we expose the full MagicDNS name `api.jay-snake.ts.net` or something that is only cluster resolvable like `api.tailscale.svc`.

**Comment by mayakacz** [comment]
This is now supported in the K8s operator: https://tailscale.com/kb/1236/kubernetes-operator/#cluster-egress

> We want to provide two-way connection between these services over our tailnet in order to avoid public ingress and authentication.

This is now documented as well: https://tailscale.com/kb/1236/kubernetes-operator/#exposing-a-service-in-one-cluster-to-another-cluster-cross-cluster-connectivity

**Comment by parasol-me** [comment]
@mayakacz Thank you for letting me know, we will definitely give this a try. It should simplify our deployments a lot!
